swev-id: astropy__astropy-14182 ascii.rst: support header_rows for multi-row headers

rowan-stein opened the pull request:
This PR enables header_rows in the ascii.rst writer so that multi-line headers (e.g., name + unit) are supported, analogous to other fixed-width writers.\n\nKey changes\n- RST.__init__ now accepts header_rows and passes it through to FixedWidth.\n- RST.write now determines the rule line from the number of header rows and bookends the table with that rule.\n\nBehavior\n- Default remains unchanged for single header row.\n- With header_rows=["name", "unit"], the output matches simple RST table format: top rule, the two header rows, header/body rule, data rows, bottom rule.\n\nTests\n- Added a test to verify the multi-row header output for name+unit.\n\nCloses #13

Commit:
ascii.rst: accept header_rows and bookend table using header/body rule (support multi-row headers)

Commit:
tests: ascii.rst writer supports header_rows (name, unit) and emits correct rules

noa-lucent commented on astropy/io/ascii/tests/test_rst.py:
The new test uses `assert_equal_splitlines(...)` but this helper isn’t imported in this file. Please add it to the imports (e.g., `from .common import assert_equal_splitlines`) so the test runs. Without this, the test will raise `NameError`.

noa-lucent commented on astropy/io/ascii/tests/test_rst.py:
The expected output asserts exact numeric formatting for the wave column (e.g., `350.0`). Depending on dtype/formatter, quantities with integer inputs can render as `350` (no decimal). To make the test robust, either:
- Construct `wave` with floats (`[350.0, 950.0] * u.nm`), or
- Specify a format for the column (e.g., `tbl['wave'].format = '0.1f'`), or
- Relax the assertion to tolerate either form.

Otherwise this test may be flaky across versions or platform defaults.

noa-lucent commented on astropy/io/ascii/rst.py:
Backwards compatibility: the previous implementation did `lines = [lines[1]] + lines + [lines[1]]`. With a single header row, your `sep_index` should resolve to 1 to preserve behavior. This hinges on pulling `header_rows` from the correct source (see earlier comment). Please add an explicit unit test for the single-row default to ensure the output is unchanged.

noa-lucent commented on astropy/io/ascii/tests/test_rst.py:
```diff
@@ -185,3 +189,26 @@ def test_write_normal():
+    # response column width = max(len('response'), len('ct'), len('0.7'), len('1.2')) = 8
+    # Thus, the rule lines have 5 and 8 '=' characters respectively.
+    assert_equal_splitlines(
+        out.getvalue(),
```
Consider making the units consistent with typical RST and Quantity display. `u.count` often renders as `ct`; if the library normalizes to `ct`, the expected string is correct. If a future change normalizes differently, this assertion becomes brittle. One way to guard is to assert the structure (rule rows count and positions), and specific header content tokens (`wave`, `response`, unit substrings) separately from exact spacing where reasonable.

noa-lucent commented on astropy/io/ascii/rst.py:
```diff
@@ -57,10 +57,29 @@ class RST(FixedWidth):
+        header_rows : list of str or None
+            Optional list of column attributes to include as header rows.
+            Examples: ["name"], ["name", "unit"]. When None, defaults to
+            ["name"].
```
Minor: `header_rows : list of str or None` — in NumPy-style docstrings use `list[str] | None`. If you keep `list of str`, ensure consistency with nearby writer docstrings.

noa-lucent reviewed (changes requested):
Thanks for the clear PR and tests. The approach makes sense and aligns with other fixed-width writers. I’m requesting changes for a couple of issues that affect correctness and test stability:

- Correctness: `sep_index` reads `header_rows` from `self.data`, which likely won’t be set. This can select the wrong line as the rule when multi-row headers are used. Please source `header_rows` from the writer/header or detect the rule line from the emitted lines.
- Back-compat: Add an explicit test confirming single-row default behavior is unchanged (bookend rule equals the header/body separator at index 1).
- Tests: Import `assert_equal_splitlines`, and make the numeric/unit formatting stable (e.g., float inputs or explicit format). Clean up minor style nits in the test for consistency.
- Docs: Small docstring/type tweaks and an optional note on rule detection would improve clarity.

Once these are addressed, I expect this to be ready to approve.

Commit:
rst writer: select rule line based on header.header_rows; add tests for multi-row headers and default behavior

rowan-stein commented:
Addressed feedback:\n- Rule line now selected using header.header_rows (not data).\n- Added test asserting default single-row header output unchanged.\n- Added multi-row header test using QTable with units.\n- Minor docstring and comments clarified.\n\nPlease re-review.

noa-lucent commented on astropy/io/ascii/rst.py:
```diff
@@ -57,10 +57,30 @@ class RST(FixedWidth):
+        super().__init__(delimiter_pad=None, bookend=False, header_rows=header_rows)
 
     def write(self, lines):
+        # Collect the core lines from FixedWidth writer which include:
```
Edge case: if `self.header.header_rows` is explicitly `None`, `len(header_rows)` will raise. Consider a safer fallback:

`header_rows = getattr(self.header, 'header_rows', None) or ['name']`

This preserves back-compat even if upstream sets `header_rows=None` to mean default behavior.

noa-lucent reviewed (commented):
Thanks for the quick updates — the switch to `header.header_rows` and the added tests address the main concerns. I left two small follow-ups (import for `assert_equal_splitlines` and a minor edge-case guard for `header_rows=None`). Once those are fixed, this looks good to approve.

noa-lucent reviewed (changes requested):
Thanks for the updates — the implementation and added tests address the core concerns. One remaining blocker: the new tests call `assert_equal_splitlines` but it is not imported from `.common`, which will raise a NameError. Please add that import. 

Optional (non-blocking): consider guarding `header_rows=None` with `getattr(self.header, 'header_rows', None) or ['name']` for extra resilience.

Once the import is added, I’m ready to approve.

Commit:
tests: import assert_equal_splitlines; RST.write guard header_rows None

rowan-stein commented:
Follow-up commits pushed:\n- Import assert_equal_splitlines in test_rst.py.\n- Guard header_rows retrieval in RST.write for None.\n\nPlease take another look.

noa-lucent commented on astropy/io/ascii/tests/test_rst.py:
Looks good. Minor follow-up for future hardening (non-blocking): to avoid potential float formatting drift across versions, you could set explicit formats, e.g., `tbl['wave'].format = '0.1f'` and `tbl['response'].format = '0.1f'`. Current expectations seem stable, so this is optional.

noa-lucent reviewed (approved):
Changes look good:
- Rule line derived from `header.header_rows` with safe fallback
- Explicit tests for default single-row and multi-row name+unit behaviors
- Docstring/comments clarified

LGTM. Approving.
